### PR TITLE
Fix Chunk.empty.toArray throws ClassCastException

### LINKED ISSUE #786: Chunk.empty.toArray throws ClassCastException
`val a = Chunk.empty.toArray[String]`
gives
`java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Ljava.lang.String;`

I'm not sure why the casting is used in the implementation.

### PR DESCRIPTION
**Pull request by wi101**
Fixes #786

### CODE REVIEW

**Comment by wi101** [comment]
@iravid sorry for the history of my commits (don't how to squash them 🙈 )

File: streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala [hunk 1]
```diff
@@ -84,6 +86,18 @@ object ChunkSpec
...
+        },
+        testM("toArray with elements of type String") {
+          check(chunkGen(stringGen)) { c =>
+            assert(c.toArray.toSeq, equalTo(c.toSeq))
+          }
+        },
+        test("toArray for a Chunk of a given type") {
+          val v: Vector[Any] = Vector("String", 1, Value(2))
+          assert(Chunk.fromIterable(v).toArray.toVector, equalTo(v))
+        },
```
Comment (iravid) [inline streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala]: Could  you add  @Jentsch 's testcase? `Chunk(1).filter(_ == 2).map(_.toString).toArray[String]`

**Comment by iravid** [comment]
@wi101 don't worry we'll get rid of that when merging

File: streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala (hunk 1 above)
Comment (wi101) [reply streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala]: Sure

Comment (wi101) [reply streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala]: it works but sounds there is a problem in circleci will check

**Comment by vasilmkd** [comment]
> @iravid sorry for the history of my commits (don't how to squash them see_no_evil )

I found this very helpful when learning git.
https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 2]
```diff
@@ -301,12 +301,6 @@ sealed trait Chunk[+A] { self =>
        else Chunk.Arr(dest))
   }
 
-  /**
-   * Materializes a chunk into a chunk backed by an array. This method can
-   * improve the performance of bulk operations.
-   */
-  def materialize[A1 >: A]: Chunk[A1] = Chunk.Arr(toArray)
```
[the remark this answers is not in the document]
Comment (wi101) [reply core/shared/src/main/scala/zio/Chunk.scala]: Yes, will do that and try to make it work

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 3]
```diff
@@ -43,7 +43,7 @@ sealed trait Chunk[+A] { self =>
    * Returns a filtered, mapped subset of the elements of this chunk.
    */
   def collect[B](p: PartialFunction[A, B]): Chunk[B] =
-    self.materialize.collect(p)
```
[the remark this answers is not in the document]
Comment (wi101) [reply core/shared/src/main/scala/zio/Chunk.scala]:
```
java.lang.StackOverflowError
	at zio.Chunk.equals(Chunk.scala:82)
	at zio.Chunk.equals$(Chunk.scala:80)
	at zio.Chunk$Empty$.equals(Chunk.scala:838)
	at zio.Chunk.toArrayOption(Chunk.scala:389)
	at zio.Chunk.materialize(Chunk.scala:308)
	at zio.Chunk.materialize$(Chunk.scala:308)
	at zio.Chunk$Empty$.materialize(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
	at zio.Chunk.collect(Chunk.scala:46)
	at zio.Chunk.collect$(Chunk.scala:45)
	at zio.Chunk$Empty$.collect(Chunk.scala:838)
```

[the remark this answers is not in the document]
Comment (wi101) [reply core/shared/src/main/scala/zio/Chunk.scala]: when I use `materialize` I got ` StackOverFlowError` , so I implemented collect using pattern match to avoid having a recursive function for collect, maybe this will solve the problem, but i don't know if materialize would be useful

[the remark this answers is not in the document]
Comment (wi101) [reply core/shared/src/main/scala/zio/Chunk.scala]: oups I find it it was because in materializer the toArrayOption should explicitly use the A1 type and not the A

[the remark this answers is not in the document]
Comment (jdegoes) [reply core/shared/src/main/scala/zio/Chunk.scala]: @wi101 

The reason this stack overflows: `self.materialize.collect(p)` will end up calling `Chunk#collect` again for the empty chunk.

If you override the method inside `Empty`, and make it return `Empty`, we can solve this prolem:

```scala
object Empty extends Chunk[Nothing] {
  ...
  override def collect[B](f: PartialFunction[A, B]): Chunk[B] = Empty
  ...
}
```

It worked before because the `materialize` would convert the `Empty` into array chunk, which has an (overridden) implementation of `collect`. But now that is not the case.

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 4]
```diff
@@ -383,11 +388,17 @@ sealed trait Chunk[+A] { self =>
...
 
   /**
-   * Converts the chunk into an array.
+   * A helper function that converts the chunk into an array if it is not empty.
    */
-  def toArray[A1 >: A]: Array[A1] = {
-    implicit val A1: ClassTag[A1] = Chunk.classTagOf(self)
+  private def toArrayOption[A1 >: A]: Option[Array[A1]] = self match {
+    case Chunk.Empty => None
+    case chunk       => Some(chunk.toArray(Chunk.classTagOf(self)))
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/Chunk.scala]: Looks great!

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 5]
```diff
@@ -408,7 +419,7 @@ sealed trait Chunk[+A] { self =>
   }
 
   override def toString: String =
-    toArray.mkString(s"${self.getClass.getSimpleName}(", ",", ")")
+    toArrayOption.fold("Array()")(_.mkString(s"${self.getClass.getSimpleName}(", ",", ")"))
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/Chunk.scala]: Replace `Array` by `${self.getClass.getSimpleName}`???

File: streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala [hunk 6]
```diff
@@ -84,6 +86,21 @@ object ChunkSpec
             assert(c.toArray.toSeq, equalTo(c.toSeq))
           }
         },
+        test("toArray for an empty Chunk of type String") {
+          assert(Chunk.empty.toArray[String], equalTo(Array.empty[String]))
```
Comment (jdegoes) [inline streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala]: 👍

File: streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala [hunk 7]
```diff
@@ -84,6 +86,21 @@ object ChunkSpec
...
+        },
+        testM("toArray with elements of type String") {
+          check(chunkGen(stringGen)) { c =>
+            assert(c.toArray.toSeq, equalTo(c.toSeq))
+          }
+        },
+        test("toArray for a Chunk of any type") {
+          val v: Vector[Any] = Vector("String", 1, Value(2))
+          assert(Chunk.fromIterable(v).toArray.toVector, equalTo(v))
+        },
```
Comment (jdegoes) [inline streams-tests/jvm/src/test/scala/zio/stream/ChunkSpec.scala]: Let's add a test of `collect` on `Empty`. 👍

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (wi101) [reply core/shared/src/main/scala/zio/Chunk.scala]: It makes 💯sense!! thanks

**Comment by wi101** [comment]
Thanks @jdegoes for your guidance :-)

**Comment by jdegoes** [comment]
@wi101 Awesome work, thank you! And good to see you back in action. 😄 💪 🎉
